File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/api/handlers/CreateGameHttpHandler.java
package systems.zlink.samples.tictactoe.server.api.handlers;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import systems.zlink.framework.spots.ZLinkSpotManager;
import systems.zlink.samples.tictactoe.server.configuration.ApiSettings;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameHttpReq;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameHttpRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayNodeInfo;
import systems.zlink.samples.tictactoe.shared.contracts.TicTacToeGameCreateReq;

import java.util.concurrent.CompletionStage;

@RestController
public final class CreateGameHttpHandler {
    private final ZLinkSpotManager spots;
    private final ApiSettings settings;

    public CreateGameHttpHandler(ZLinkSpotManager spots, ApiSettings settings) {
        this.spots = spots;
        this.settings = settings;
    }

    @PostMapping("/games")
    public CompletionStage<CreateGameHttpRes> handle(@RequestBody CreateGameHttpReq request) {
        String gameName = gameName(request);
        // --8<-- [start:doc-create]
        return spots.create("tictactoe.game") // 이 stable type을 등록한 node가 후보가 된다.
                .inMesh(SampleNames.SpotMesh) // Spot을 만들 mesh를 고른다.
                .request(new TicTacToeGameCreateReq(gameName, SampleNames.RequiredLevel))
                .timeout(SampleNames.RequestTimeout)
                .submit() // Java의 비동기 완료 terminal이다.
                // --8<-- [end:doc-create]
                .thenApply(
                        created ->
                                new CreateGameHttpRes(
                                        created.spot().spotId(),
                                        gameName,
                                        settings.playEndpoints(),
                                        settings.playEndpoints().stream()
                                                .map(PlayNodeInfo::new)
                                                .toList(),
                                        SampleNames.RequiredLevel));
    }

    private static String gameName(CreateGameHttpReq request) {
        return request.gameName() == null || request.gameName().isBlank()
                ? "tictactoe-game"
                : request.gameName();
    }
}

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/api/handlers/CreateGameHttpHandler.kt
package systems.zlink.samples.kotlin.tictactoe.server.api.handlers

import kotlinx.coroutines.future.await
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import systems.zlink.framework.spots.ZLinkSpotManager
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameHttpReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameHttpRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayNodeInfo
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.TicTacToeGameCreateReq

@RestController
class CreateGameHttpHandler(
    private val spots: ZLinkSpotManager,
    private val settings: SampleSettings,
) {
    @PostMapping("/games")
    suspend fun handle(@RequestBody request: CreateGameHttpReq): CreateGameHttpRes {
        val gameName = request.gameName?.takeIf { it.isNotBlank() } ?: "tictactoe-game"
        // --8<-- [start:doc-create]
        val created =
            spots
                .create("tictactoe.game") // 이 stable type을 등록한 node가 후보가 된다.
                .inMesh(SampleNames.SpotMesh) // Spot을 만들 mesh를 고른다.
                .request(TicTacToeGameCreateReq(gameName, SampleNames.RequiredLevel))
                .timeout(SampleNames.RequestTimeout)
                .submit()
                .await() // Kotlin의 비동기 완료 terminal이다.
        // --8<-- [end:doc-create]
        return CreateGameHttpRes(
            roomId = created.spot.spotId,
            gameName = gameName,
            playEndpoints = settings.playEndpoints,
            playNodes = settings.playEndpoints.map(::PlayNodeInfo),
            requiredLevel = SampleNames.RequiredLevel,
        )
    }
}
